exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,417 @@
+        row->array->data[i].type = cString;
+        row->array->data[i].string = string_createCString(values[i]);
+    }
+    cell_arrayAppendElementPointer(rowscell, row);
+    return 0;
+}
+
+
+
+void sqlite_open(TExecutor *exec)
+{
+    char *name = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    sqlite3 *db;
+    int r = sqlite3_open(name, &db);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db), number_from_sint64(r));
+        free(name);
+        return;
+    }
+
+    push(exec->stack, cell_fromObject(object_createDatabaseObject(db)));
+    free(name);
+}
+
+void sqlite_exec(TExecutor *exec)
+{
+    Cell *parameters = cell_fromCell(top(exec->stack)); pop(exec->stack);
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Cell *pdb = cell_fromCell(top(exec->stack)); pop(exec->stack);
+
+    sqlite3_stmt *stmt;
+    Object *db = check_database(exec, pdb->object);
+    if (db == NULL) {
+        goto cleanup;
+    }
+    Cell *rows = cell_createArrayCell(0);
+    int r = sqlite3_prepare_v2(db->ptr, sql, -1, &stmt, NULL);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+        goto cleanup;
+    }
+    for (int i = 0; i < parameters->dictionary->len; i++) {
+        int c = sqlite3_bind_parameter_index(stmt, TCSTR(parameters->dictionary->data[i].key));
+        if (c == 0) {
+            exec->rtl_raise(exec, "SqlException.ParameterName", TCSTR(parameters->dictionary->data[i].key), BID_ZERO);
+            goto cleanup;
+        }
+        r = sqlite3_bind_text(stmt, c, TCSTR(parameters->dictionary->data[i].value->string), -1, SQLITE_TRANSIENT);
+        if (r != SQLITE_OK) {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto cleanup;
+        }
+    }
+    int columns = sqlite3_column_count(stmt);
+    for (;;) {
+        r = sqlite3_step(stmt);
+        if (r == SQLITE_DONE) {
+            break;
+        }
+        if (r == SQLITE_ROW) {
+            Cell *row = cell_createArrayCell(0);
+            for (int i = 0; i < columns; i++) {
+                cell_arrayAppendElementPointer(row, cell_fromCString((const char *)sqlite3_column_text(stmt, i)));
+            }
+            cell_arrayAppendElementPointer(rows, row);
+        } else {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto cleanup;
+        }
+    }
+    push(exec->stack, rows);
+
+cleanup:
+    sqlite3_finalize(stmt);
+    free(sql);
+    cell_freeCell(parameters);
+    cell_freeCell(pdb);
+}
+
+void sqlite_execOne(TExecutor *exec)
+{
+    Cell *parameters = cell_fromCell(top(exec->stack)); pop(exec->stack);
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Cell *pdb = cell_fromCell(top(exec->stack)); pop(exec->stack);
```

Comment:
Do you really need to make a *copy* of the cell here? I can't even see where you free this in the normal case.

Reply:
It's not MANDATORY that I make a copy, but if I don't, then the code starts to look weird.  Like.
`Cell *pdb = top(exec->stack);
Object *db = check_database(exec, pdb->object);
pop(exec->stack);
if (db == NULL)
...
`
So, I just opted to copy the cell.  I'm not thrilled with either solution, so I'll take your direction on which method you think is best.  Based on this suggestion, I'd like to go back over all the instances that use this style of construction and redo them, so the code is the same in all places.

Reply:
Actually.  Now that I look at what I wrote above, I think I'm going to attack it from a different perspective all together.  I'm going to eliminate the Cell altogether.  Don't really need it.

Reply:
Fixed.  Used a new approach.